File: slide-rs/src/recv.rs
use anyhow::{bail, Result};
use console::style;
use indicatif::{ProgressBar, ProgressStyle};
use std::fs;
use std::io::Write;
use std::path::Path;
use std::thread;
use std::time::{Duration, Instant};

use crate::protocol::*;

pub fn recv_session(port_name: &str, baud: u32, output_dir: &str, debug: bool) -> Result<()> {
    println!(
        "{} v0.2 — Serial Line Inter-Device Exchange",
        style("SLIDE").cyan().bold()
    );
    println!("  Port:   {} @ {} baud", style(port_name).yellow(), baud);
    println!("  Mode:   {}", style("receive").green().bold());
    println!("  Output: {}", output_dir);
    println!();

    fs::create_dir_all(output_dir)?;
    let mut port = open_serial(port_name, baud)?;

    // Handshake: wait for sender's RDY, echo back
    let spin = ProgressBar::new_spinner();
    spin.set_style(
        ProgressStyle::default_spinner()
            .template("{spinner:.cyan} {msg}")
            .unwrap(),
    );
    spin.set_message("Waiting for Z80 sender (start SLIDE S <file> on Z80 now)...");
    spin.enable_steady_tick(Duration::from_millis(100));

    let deadline = Instant::now() + Duration::from_secs(60);
    loop {
        let remaining = deadline.saturating_duration_since(Instant::now());
        if remaining.is_zero() {
            spin.abandon_with_message(style("Timeout waiting for Z80").red().to_string());
            bail!("Timeout waiting for Z80 ready signal");
        }
        match recv_control(port.as_mut(), remaining.min(Duration::from_secs(2))) {
            Ok(Control::Rdy) => break,
            _ => continue,
        }
    }

    // Echo RDY back
    port.write_all(&[CTRL_RDY])?;
    port.flush()?;
    spin.finish_with_message(format!("{} Z80 connected.", style("✓").green().bold()));
    thread::sleep(Duration::from_millis(50));
    port.clear(serialport::ClearBuffer::Input)?;

    // Receive files until FIN
    let mut file_count = 0u32;
    let session_start = Instant::now();
    let mut total_bytes = 0u64;

    loop {
        println!(
            "\n{}",
            style(format!("── Waiting for file {}... ──", file_count + 1)).dim()
        );
        match recv_one_file(port.as_mut(), output_dir, debug)? {
            RecvResult::Ok(bytes) => {
                file_count += 1;
                total_bytes += bytes;
            }
            RecvResult::Fin => {
                send_control(port.as_mut(), CTRL_FIN, None)?;
                if debug {
                    eprintln!("  DEBUG sent FIN echo");
                }
                break;
            }
            RecvResult::Error => {
                println!("  File transfer failed, aborting session.");
                break;
            }
        }
    }

    let elapsed = session_start.elapsed();
    println!();
    println!(
        "{} {} file(s) received, {} bytes in {:.1}s",
        style("✓").green().bold(),
        file_count,
        total_bytes,
        elapsed.as_secs_f64()
    );

    Ok(())
}

enum RecvResult {
    Ok(u64),
    Fin,
    Error,
}

fn recv_one_file(port: &mut dyn serialport::SerialPort, output_dir: &str, debug: bool) -> Result<RecvResult> {
    // Receive header frame (seq 0)
    let (filename, filesize) = match recv_frame(port, Duration::from_secs(30))? {
        FrameResult::Fin => return Ok(RecvResult::Fin),
        FrameResult::Data(frame) => {
            if debug {
                eprintln!(
                    "  DEBUG header seq={} len={}: {:02x?}",
                    frame.seq,
                    frame.payload.len(),
                    &frame.payload
                );
            }
            parse_header(&frame.payload)?
        }
    };

    let filepath = Path::new(output_dir).join(&filename);
    println!(
        "  {} — {} bytes",
        style(&filename).bold(),
        filesize
    );

    // ACK header
    send_control(port, CTRL_ACK, Some(0))?;

    // Progress bar
    let pb = ProgressBar::new(filesize as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("  {bar:50.green/dim} {bytes}/{total_bytes}  {msg}")
            .unwrap()
            .progress_chars("━╸─"),
    );

    // Receive data frames
    let mut expected_seq: u8 = 1;
    let mut received_data: Vec<u8> = Vec::with_capacity(filesize as usize);
    let mut retry_count = 0u32;
    let max_retries = 15;

    loop {
        let frame_result = recv_frame(port, Duration::from_secs(10));

        match frame_result {
            Err(e) => {
                retry_count += 1;
                if retry_count >= max_retries {
                    pb.abandon_with_message(style("too many retries").red().to_string());
                    return Ok(RecvResult::Error);
                }
                if debug {
                    eprintln!("  DEBUG error: {e}, NAK seq={expected_seq}");
                }
                pb.set_message(
                    style(format!("retry {retry_count}/{max_retries}"))
                        .yellow()
                        .to_string(),
                );
                send_control(port, CTRL_NAK, Some(expected_seq))?;
                continue;
            }
            Ok(FrameResult::Fin) => {
                pb.abandon_with_message(style("unexpected FIN").red().to_string());
                return Ok(RecvResult::Error);
            }
            Ok(FrameResult::Data(frame)) => {
                retry_count = 0;

                // EOF = zero-length payload
                if frame.payload.is_empty() {
                    if debug {
                        eprintln!("  DEBUG EOF frame seq={}", frame.seq);
                    }
                    send_control(port, CTRL_ACK, Some(expected_seq))?;
                    break;
                }

                // Sequence check
                if frame.seq != expected_seq {
                    if debug {
                        eprintln!(
                            "  DEBUG seq mismatch: got {} expected {}",
                            frame.seq, expected_seq
                        );
                    }
                    send_control(port, CTRL_NAK, Some(expected_seq))?;
                    continue;
                }

                received_data.extend_from_slice(&frame.payload);
                expected_seq = expected_seq.wrapping_add(1);

                if debug {
                    eprintln!(
                        "  DEBUG frame seq={} len={} total={}",
                        frame.seq,
                        frame.payload.len(),
                        received_data.len()
                    );
                }

                // ACK every WIN_SIZE frames
                if (expected_seq.wrapping_sub(1)) & (WIN_SIZE as u8 - 1) == 0 {
                    send_control(port, CTRL_ACK, Some(frame.seq))?;
                    if debug {
                        eprintln!("  DEBUG ACK seq={}", frame.seq);
                    }
                }

                pb.set_position(received_data.len().min(filesize as usize) as u64);
            }
        }
    }

    // Truncate to actual size (remove CP/M padding)
    let file_data = if received_data.len() > filesize as usize {
        &received_data[..filesize as usize]
    } else {
        &received_data
    };

    fs::write(&filepath, file_data)?;

    pb.finish_with_message(format!(
        "saved {}",
        style(filepath.display()).dim()
    ));

    Ok(RecvResult::Ok(filesize as u64))
}

fn parse_header(payload: &[u8]) -> Result<(String, u32)> {
    let null_idx = payload
        .iter()
        .position(|&b| b == 0)
        .ok_or_else(|| anyhow::anyhow!("No null terminator in header"))?;
    let filename = String::from_utf8(payload[..null_idx].to_vec())?;
    if payload.len() < null_idx + 5 {
        bail!("Header too short for file size");
    }
    let filesize = u32::from_le_bytes(payload[null_idx + 1..null_idx + 5].try_into()?);
    Ok((filename, filesize))
}

File: slide-py/slide/recv.py
#!/usr/bin/env python3
"""
SLIDE - Serial Line Inter-Device (file) Exchange
PC-side receiver: accepts files sent from Z80/CP/M.

Usage: slide-recv <serial_port> [--baud 19200] [--output-dir .]
"""

import struct
import sys
import os
import time
import argparse
from pathlib import Path

from slide.common import (
    SOF, CTRL_ACK, CTRL_NAK, CTRL_RDY, CTRL_FIN, CTRL_CAN,
    WIN_SIZE, FRAME_SIZE,
    crc16_ccitt, open_serial,
)


def recv_frame(ser, timeout: float = 10.0):
    """
    Receive a SLIDE frame from the serial port.
    Returns: (seq, payload) on success
    Raises: TimeoutError, ValueError (CRC mismatch)
    """
    ser.timeout = timeout

    # Wait for SOF
    while True:
        b = ser.read(1)
        if not b:
            raise TimeoutError("Timeout waiting for SOF")
        if b[0] == SOF:
            break
        if b[0] == CTRL_FIN:
            raise _FinReceived()

    return _recv_frame_after_sof(ser)


class _FinReceived(Exception):
    """Raised when FIN is received instead of SOF."""
    pass


def _recv_frame_after_sof(ser):
    """Receive the rest of a frame after SOF has been consumed."""
    # SEQ
    b = ser.read(1)
    if not b:
        raise TimeoutError("Timeout waiting for SEQ")
    seq = b[0]

    # LEN_H, LEN_L
    b = ser.read(2)
    if len(b) < 2:
        raise TimeoutError("Timeout waiting for LEN")
    length = (b[0] << 8) | b[1]

    # Payload
    payload = b''
    if length > 0:
        payload = ser.read(length)
        if len(payload) < length:
            raise TimeoutError(f"Timeout in payload: got {len(payload)}/{length}")

    # CRC_H, CRC_L
    b = ser.read(2)
    if len(b) < 2:
        raise TimeoutError("Timeout waiting for CRC")
    rx_crc = (b[0] << 8) | b[1]

    # Verify CRC over SEQ + LEN + PAYLOAD
    crc_data = bytes([seq, length >> 8, length & 0xFF]) + payload
    calc_crc = crc16_ccitt(crc_data)
    if calc_crc != rx_crc:
        raise ValueError(f"CRC mismatch: calc=0x{calc_crc:04X} rx=0x{rx_crc:04X}")

    return (seq, payload)


def send_control(ser, ctrl: int, seq: int = None):
    """Send a control byte, optionally followed by a sequence number."""
    ser.write(bytes([ctrl]))
    if seq is not None:
        ser.write(bytes([seq]))
    ser.flush()


def recv_one_file(ser, output_dir: str, debug: bool = False) -> bool:
    """
    Receive a single file (header already expected as next frame).
    Returns True on success, False on error.
    """
    # --- Receive header frame (seq 0) ---
    try:
        seq, payload = recv_frame(ser, timeout=30.0)
    except _FinReceived:
        return None  # session done
    except (TimeoutError, ValueError) as e:
        print(f"  Error receiving header: {e}")
        return False

    if debug:
        print(f"  DEBUG header seq={seq} len={len(payload)}: {payload.hex(' ')}")

    # Parse header: null-terminated filename + 4-byte LE size
    null_idx = payload.index(0)
    filename = payload[:null_idx].decode('ascii')
    filesize = struct.unpack('<I', payload[null_idx+1:null_idx+5])[0]

    filepath = os.path.join(output_dir, filename)
    print(f"  Receiving: {filename} ({filesize} bytes)")

    # ACK the header
    send_control(ser, CTRL_ACK, 0)

    # --- Receive data frames ---
    expected_seq = 1
    received_data = bytearray()
    retry_count = 0
    max_retries = 15

    while True:
        try:
            seq, payload = recv_frame(ser, timeout=10.0)
        except _FinReceived:
            print("  Unexpected FIN during file transfer")
            return False
        except TimeoutError as e:
            retry_count += 1
            if retry_count >= max_retries:
                print(f"\n  Aborted: too many retries")
                return False
            if debug:
                print(f"  DEBUG timeout, NAK seq={expected_seq}")
            send_control(ser, CTRL_NAK, expected_seq)
            continue
        except ValueError as e:
            retry_count += 1
            if retry_count >= max_retries:
                print(f"\n  Aborted: too many CRC errors")
                return False
            if debug:
                print(f"  DEBUG CRC error: {e}, NAK seq={expected_seq}")
            send_control(ser, CTRL_NAK, expected_seq)
            continue

        # Reset retry counter on successful frame
        retry_count = 0

        # Zero-length = end of file
        if len(payload) == 0:
            if debug:
                print(f"  DEBUG EOF frame seq={seq}")
            send_control(ser, CTRL_ACK, expected_seq)
            break

        # Check sequence
        if seq != expected_seq:
            if debug:
                print(f"  DEBUG seq mismatch: got {seq} expected {expected_seq}")
            send_control(ser, CTRL_NAK, expected_seq)
            continue

        received_data.extend(payload)
        expected_seq = (expected_seq + 1) & 0xFF

        if debug:
            print(f"  DEBUG frame seq={seq} len={len(payload)} total={len(received_data)}")

        # ACK every WIN_SIZE frames
        if (expected_seq - 1) & (WIN_SIZE - 1) == 0:
            send_control(ser, CTRL_ACK, seq)
            if debug:
                print(f"  DEBUG ACK seq={seq}")

        # Progress
        pct = min(len(received_data) * 100 // filesize, 100) if filesize > 0 else 100
        bar = '#' * (pct // 2) + '-' * (50 - pct // 2)
        print(f"\r  [{bar}] {pct}%", end='')

    # Truncate to actual file size (remove CP/M padding)
    file_data = bytes(received_data[:filesize])

    with open(filepath, 'wb') as f:
        f.write(file_data)

    print(f"\n  Saved: {filepath} ({len(file_data)} bytes)")
    return True


def recv_session(port: str, baud: int = 19200, output_dir: str = '.', debug: bool = False):
    """Receive one or more files from Z80 using SLIDE v0.2 session protocol."""

    print(f"SLIDE v0.2 - Serial Line Inter-Device Exchange (receive)")
    print(f"  Port: {port} @ {baud} baud")
    print(f"  Output: {output_dir}")
    print()

    os.makedirs(output_dir, exist_ok=True)
    ser = open_serial(port, baud)

    # --- Handshake: wait for sender's RDY, echo RDY back ---
    print("Waiting for Z80 sender (start SLIDE S <file> on Z80 now)...")
    ser.timeout = 60
    while True:
        b = ser.read(1)
        if not b:
            print("ERROR: Timeout waiting for Z80 ready signal")
            ser.close()
            return
        if b[0] == CTRL_RDY:
            break

    # Echo RDY back
    ser.write(bytes([CTRL_RDY]))
    ser.flush()
    print("Z80 connected.")
    time.sleep(0.05)
    ser.reset_input_buffer()

    # --- Receive files until FIN ---
    file_count = 0
    start_time = time.time()

    while True:
        print(f"\n--- Waiting for file {file_count + 1}... ---")
        result = recv_one_file(ser, output_dir, debug)
        if result is None:
            # FIN received — session done
            send_control(ser, CTRL_FIN)
            if debug:
                print("  DEBUG sent FIN echo")
            break
        elif result:
            file_count += 1
        else:
            print("  File transfer failed, aborting session.")
            break

    elapsed = time.time() - start_time
    print(f"\nSession complete — {file_count} file(s) received in {elapsed:.1f}s.")
    ser.close()


def main():
    parser = argparse.ArgumentParser(description='SLIDE - Receive files from Z80/CP/M')
    parser.add_argument('port', help='Serial port (e.g., /dev/ttyUSB0, COM3)')
    parser.add_argument('--baud', type=int, default=19200, help='Baud rate (default: 19200)')
    parser.add_argument('--output-dir', default='.', help='Directory for received files (default: .)')
    parser.add_argument('--debug', action='store_true', help='Show wire-level debug output')
    args = parser.parse_args()

    recv_session(args.port, args.baud, args.output_dir, args.debug)


if __name__ == '__main__':
    main()
